packages/core/src/routes/experience/classes/verifications/password-verification.ts

```diff
@@ -0,0 +1,94 @@
+import {
+  VerificationType,
+  interactionIdentifierGuard,
+  type InteractionIdentifier,
+} from '@logto/schemas';
+import { type ToZodObject } from '@logto/schemas/lib/utils/zod.js';
+import { generateStandardId } from '@logto/shared';
+import { z } from 'zod';
+
+import RequestError from '#src/errors/RequestError/index.js';
+import type Libraries from '#src/tenants/Libraries.js';
+import type Queries from '#src/tenants/Queries.js';
+import assertThat from '#src/utils/assert-that.js';
+
+import { findUserByIdentifier } from '../../utils.js';
+
+import { type Verification } from './verification.js';
+
+export type PasswordVerificationRecordData = {
+  id: string;
+  type: VerificationType.Password;
+  identifier: InteractionIdentifier;
+  /* The userId of the user that has been verified */
+  userId?: string;
+};
+
+export const passwordVerificationRecordDataGuard = z.object({
+  id: z.string(),
+  type: z.literal(VerificationType.Password),
+  identifier: interactionIdentifierGuard,
+  userId: z.string().optional(),
+}) satisfies ToZodObject<PasswordVerificationRecordData>;
+
+/**
+ * PasswordVerification is a verification record that verifies a user's identity
+ * using identifier and password
+ */
+export class PasswordVerification implements Verification {
```

Comment:
will we add unit tests to this class?

Reply:
Implement the integration test to cover all these for now. Will add some UT after the full implementation.